### PR TITLE
Update base_retriever.py

### PR DESCRIPTION
**Pull request by danerlt**
Update base_retriever.py _handle_recursive_retrieval deduplication algorithm

# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-core/llama_index/core/base/base_retriever.py [hunk 1]
```diff
@@ -166,12 +166,12 @@ def _handle_recursive_retrieval(
...
-            n
-            for n in retrieved_nodes
-            if not (n.node.hash in seen or seen.add(n.node.hash))  # type: ignore[func-returns-value]
-        ]
+        res = dict()
+        for n in retrieved_nodes:
+            node_hash = n.node.hash
+            if node_hash not in res:
+                res[node_hash] = n
+        return list(res.values()]
```
Comment (logan-markewich) [inline llama-index-core/llama_index/core/base/base_retriever.py:174]: can you provide some motivation for this change? Using a `set()` like before should be just fine

Was this tested? I see a stray bracket at the end of the line here too

Comment (danerlt) [reply llama-index-core/llama_index/core/base/base_retriever.py:174]: I think using a dictionary is more readable and simpler than using a set

Comment (danerlt) [reply llama-index-core/llama_index/core/base/base_retriever.py:174]: > can you provide some motivation for this change? Using a `set()` like before should be just fine
> 
> Was this tested? I see a stray bracket at the end of the line here too

You are right, there is indeed an extra bracket at the end

Comment (logan-markewich) [reply llama-index-core/llama_index/core/base/base_retriever.py:174]: .... can you remove it? :)

**Comment by logan-markewich** [comment]
Going to close this one out for now, its quite outdated at this point

If its still an issue, please raise an updated PR

Sorry for letting this one sit for so long!

**Pull request closed without merging**